jbi/queue.py

```diff
@@ -0,0 +1,207 @@
+
+    async def remove(self, bug_id: int, identifier: str):
+        path = self.location / f"{bug_id}" / (identifier + ".json")
+        os.remove(path)
+
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        folder = self.location / f"{bug_id}"
+        results = []
+        for filepath in folder.glob("*.json"):
+            with open(folder / filepath) as f:
+                results.append(QueueItem(**json.load(f)))
+        return results
+
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        all_items = []
+        for filepath in self.location.rglob("*.json"):
+            with open(self.location / filepath) as f:
+                all_items.append(QueueItem(**json.load(f)))
+        by_bug = itertools.groupby(all_items, key=lambda i: i.payload.bug.id)
+        return {k: list(v) for k, v in by_bug}
+
+    async def size(self) -> int:
+        return sum(1 for _ in self.location.rglob("*.json"))
+
+
+class InvalidQueueDSNError(Exception):
+    pass
+
+
+class DeadLetterQueue:
+    backend: QueueBackend
+
+    def __init__(self, dsn: str):
+        parsed = urlparse(url=dsn)
+        if parsed.scheme == "memory":
+            self.backend = MemoryBackend()
+        elif parsed.scheme == "file":
+            self.backend = FileBackend(parsed.path)
+        else:
+            raise InvalidQueueDSNError(f"{parsed.scheme} is not supported")
+
+    async def postpone(self, payload: bugzilla.WebhookRequest):
+        """
+        Postpone the specified request for later.
+        """
+        item = QueueItem(payload=payload)
+        await self.backend.put(item)
+
+    async def track_failed(self, payload: bugzilla.WebhookRequest, exc: Exception):
+        """
+        Store the specified payload and exception information into the queue.
+        """
+        item = QueueItem(
+            payload=payload,
+            error=PythonException.from_exc(exc),
+        )
+        await self.backend.put(item)
+
+    async def is_blocked(self, payload: bugzilla.WebhookRequest) -> bool:
+        """
+        Return `True` if the specified `payload` is blocked and should be
+        queued instead of being processed.
+        """
+        existing = await self.backend.get(payload.bug.id)
+        return len(existing) > 0
+
+    async def retrieve(self) -> list[QueueItem]:
+        """
+        Returns the whole list of items in the queue, grouped by bug, and
+        sorted from oldest to newest.
+        """
+        existing = await self.backend.get_all()
+        # Sort by event datetime ascending.
+        sorted_existing = [
+            sorted(v, key=lambda i: i.payload.event.time or 0)
+            for v in existing.values()
+            if len(v) > 0
+        ]
+        by_oldest_bug_events = sorted(
+            sorted_existing, key=lambda items: items[0].payload.event.time or 0
+        )
+        return list(itertools.chain(*by_oldest_bug_events))
```

Comment:
After thinking about this, we shouldn't flatten this list. But if the scheduler fails to process one request, it will want to jump to the next bug. 

```python

for items in all_items:
    for item in items:
        try:
            execute_action(item.payload, actions)
        except IgnoreInvalidRequestError:
            await queue.done(item)
        except Exception as exc:
            # Jump to next bug
            break
        else:
            await queue.done(item)
```